File: ios/LaneShadow/Views/Atoms/LSMapCameraController.swift
import CoreGraphics
import SwiftUI

@MainActor
@Observable
public final class LSMapCameraController {
    public var zoomLevel: Double = 12
    public private(set) var appliedZoomDeltas: [Double] = []
    public private(set) var recenterRequestCount: Int = 0
    public private(set) var handledRecenterRequestCount: Int = 0
    public private(set) var lastRecenterOutcome: RecenterOutcome = .idle

    public enum RecenterOutcome: String, Sendable {
        case idle
        case requested
        case applied
        case unavailableNoUserLocation = "unavailable-no-user-location"
    }

    public init(zoomLevel: Double = 12) {
        self.zoomLevel = zoomLevel
    }

    public func zoomIn(step: Double = 1) {
        zoomLevel += step
    }

    public func zoomOut(step: Double = 1) {
        zoomLevel -= step
    }

    public func recenterToUserLocation() {
        recenterRequestCount += 1
        lastRecenterOutcome = .requested
    }

    public func recordAppliedZoomDelta(_ delta: Double) {
        guard delta != 0 else {
            return
        }

        appliedZoomDeltas.append(delta)
    }

    @discardableResult
    public func consumePendingRecenterRequest(outcome: RecenterOutcome) -> Bool {
        guard handledRecenterRequestCount < recenterRequestCount else {
            return false
        }

        handledRecenterRequestCount = recenterRequestCount
        lastRecenterOutcome = outcome
        return true
    }

    public var debugAccessibilityValue: String {
        let zoomDeltas = appliedZoomDeltas
            .map(Self.formatDelta)
            .joined(separator: ",")
        return "zoom-deltas=\(zoomDeltas);recenter=\(handledRecenterRequestCount)/\(recenterRequestCount);outcome=\(lastRecenterOutcome.rawValue)"
    }

    private static func formatDelta(_ delta: Double) -> String {
        if delta == floor(delta) {
            return String(format: "%.0f", delta)
        }

        return String(delta)
    }
}

File: android/app/src/main/java/com/laneshadow/ui/atoms/LSMapCameraController.kt
package com.laneshadow.ui.atoms

import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue

/**
 * LSMapCameraController — Kotlin Compose equivalent of iOS LSMapCameraController.
 *
 * Manages zoom level and recenter-to-user-location requests, tracking applied deltas
 * for instrumentation and testing. Exposes a debugAccessibilityValue for test inspection.
 */
class LSMapCameraController(
    initialZoom: Double = 12.0,
) {
    var zoomLevel by mutableStateOf(initialZoom)
        private set

    private val _appliedZoomDeltas = mutableListOf<Double>()
    val appliedZoomDeltas: List<Double> get() = _appliedZoomDeltas.toList()

    var recenterRequestCount by mutableStateOf(0)
        private set

    var handledRecenterRequestCount by mutableStateOf(0)
        private set

    var lastRecenterOutcome by mutableStateOf(RecenterOutcome.Idle)
        private set

    fun zoomIn(step: Double = 1.0) {
        zoomLevel += step
    }

    fun zoomOut(step: Double = 1.0) {
        zoomLevel -= step
    }

    fun recenterToUserLocation() {
        recenterRequestCount += 1
        lastRecenterOutcome = RecenterOutcome.Requested
    }

    fun recordAppliedZoomDelta(delta: Double) {
        if (delta != 0.0) {
            _appliedZoomDeltas.add(delta)
        }
    }

    fun consumePendingRecenterRequest(outcome: RecenterOutcome): Boolean {
        if (handledRecenterRequestCount >= recenterRequestCount) {
            return false
        }

        handledRecenterRequestCount = recenterRequestCount
        lastRecenterOutcome = outcome
        return true
    }

    val debugAccessibilityValue: String
        get() {
            val zoomDeltas = _appliedZoomDeltas
                .map { formatDelta(it) }
                .joinToString(",")
            return "zoom-deltas=$zoomDeltas;recenter=$handledRecenterRequestCount/$recenterRequestCount;outcome=${lastRecenterOutcome.rawValue}"
        }

    companion object {
        private fun formatDelta(delta: Double): String {
            return if (delta == delta.toLong().toDouble()) {
                String.format("%.0f", delta)
            } else {
                delta.toString()
            }
        }
    }
}

enum class RecenterOutcome(val rawValue: String) {
    Idle("idle"),
    Requested("requested"),
    Applied("applied"),
    UnavailableNoUserLocation("unavailable-no-user-location"),
}
